lib/plug/conn.ex

```diff
@@ -1328,50 +1340,149 @@ defmodule Plug.Conn do
+  or consider explicitly escaping the cookie value by using a function
+  such as `Base.encode64(value, padding: false)` when writing and
+  `Base.decode64(encoded, padding: false)` when reading the cookie.
+  It is important for padding to be disabled since `=` is not a valid
+  character in cookie values.
+
+  ## Signing and encrypting cookies
+
+  This function allows you to automatically sign and encrypt cookies.
+  When signing or encryption is enabled, then any Elixir value can be
+  stored in the cookie (except anonymous functions for security reasons).
+  Once a value is signed or encrypted, you must also call `fetch_cookie/2`
+  with the cookies that are either signed or encrypted.
+
+  To sign, you would do:
+
+      put_resp_cookie(conn, "my-cookie", %{user_id: user.id}, sign: true)
+
+  and then:
+
+      fetch_cookie(conn, signed: ~w(my-cookie))
+
+  To encrypt, you would do:
+
+      put_resp_cookie(conn, "my-cookie", %{user_id: user.id}, encrypt: true)
+
+  and then:
+
+      fetch_cookie(conn, encrypted: ~w(my-cookie))
+
+  By default a signed or encrypted cookie is only valid for a day, unless
+  a `:max_age` is specified.
 
   ## Options
 
     * `:domain` - the domain the cookie applies to
     * `:max_age` - the cookie max-age, in seconds. Providing a value for this
-      option will set both the _max-age_ and _expires_ cookie attributes
+      option will set both the _max-age_ and _expires_ cookie attributes.
     * `:path` - the path the cookie applies to
     * `:http_only` - when `false`, the cookie is accessible beyond HTTP
     * `:secure` - if the cookie must be sent only over https. Defaults
       to true when the connection is HTTPS
     * `:extra` - string to append to cookie. Use this to take advantage of
       non-standard cookie attributes.
+    * `:signed` - when true, signs the cookie
+    * `:encrypted` - when true, encrypts the cookie
 
   """
   @spec put_resp_cookie(t, binary, binary, Keyword.t()) :: t
   def put_resp_cookie(%Conn{} = conn, key, value, opts \\ [])
-      when is_binary(key) and is_binary(value) and is_list(opts) do
+      when is_binary(key) and is_list(opts) do
     %{resp_cookies: resp_cookies, scheme: scheme} = conn
-    cookie = [{:value, value} | opts] |> :maps.from_list() |> maybe_secure_cookie(scheme)
+    {to_send_value, opts} = maybe_sign_or_encrypt_cookie(conn, key, value, opts)
+    cookie = [{:value, to_send_value} | opts] |> Map.new() |> maybe_secure_cookie(scheme)
     resp_cookies = Map.put(resp_cookies, key, cookie)
     update_cookies(%{conn | resp_cookies: resp_cookies}, &Map.put(&1, key, value))
   end
 
+  defp maybe_sign_or_encrypt_cookie(conn, key, value, opts) do
+    {sign?, opts} = Keyword.pop(opts, :sign, false)
+    {encrypt?, opts} = Keyword.pop(opts, :encrypt, false)
+
+    case {sign?, encrypt?} do
+      {true, true} ->
+        raise ArgumentError, "only :sign or :encrypt may be given at the same time"
```

Comment:
Maybe the 'encrypted' option should be named 'signed_and_encrypted', because that's basically what happens. Anyone who learned a bit of basic crypto will know that encryption alone is not enough to ensure integrity. So they will want to do both, and might be surprised to get this error when they request just that

Reply:
Good call. I will change the error message to say something like: ":encrypt automatically implies :sign. Please pass only one or the other".

Reply:
That’s a good point. I will provide a proper support for this on Plug.Crypto.